SLP indexer and gRPC server updates

jcramer opened the pull request:
This PR provides SLP indexing with an API exposed using the gRPC server per https://docs.google.com/document/d/1pro3m0kz9b1L3iUdnXqI9wbrwB2mE8EBvZfZg94F1o4/edit?usp=sharing.

The final API is described by: 
https://github.com/simpleledgerinc/bchd/blob/slp-index/bchrpc/bchrpc.proto

zquestz commented:
This looks like a great start. Excited to see how this progresses. Let me know when you are ready for some testers!

jcramer commented:
Thanks will let you know, will probably be ready sometime monday to try it out.  Still working through some bugs atm.

jcramer commented:
This branch is ready to test.

**All SLP validation tests are passing!!**

- [Unit test](https://github.com/simpleledgerinc/bchd/blob/slp-index/blockchain/indexers/slpindex_test.go#L19) passing all [OP_RETURN parsing](https://github.com/simpleledger/slp-unit-test-data#part-a-parsing-of-op_return-scripts) tests
- [Unit test]() passing all [Input conditions](https://github.com/simpleledger/slp-unit-test-data#part-b-transaction-input-tests) tests
- Fully synchronized with [slp-comp](https://github.com/simpleledgerinc/slp-comps) mainnet and matches all validity judgements from SLPDB and gs++ indexers

I still need to hook up the parsing fuzzer and finish up gRPC mempool SLP validity.  The indexer validation only contains blocks, so mempool validation will happen on the fly probably with some caching, still need to look into it more.  After that, this will be ready for merging.  I estimate 1 more week from me.

jcramer commented:
Latest updates will return `pb.TokenMetadata` with the current mint baton location and also NFT group ID.  Just a few items remaining before we can do a final code review and merge!

cpacia commented on chaincfg/params.go:
```diff
@@ -227,6 +227,11 @@ type Params struct {
+
+	// SLP indexer parameters
+	SlpIndexStartHeight uint32
+	SlpIndexStartHash   []byte
```
Let's use a `chainhash.Hash` for this instead of a byte slice. 

It can be initialized using `newHashFromStr()`

jcramer replied:
done.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+
+// TestSlpMesesageUnitTests downloads SLP parser unit tests and checks the parser throws for each test where code != nil
+func TestSlpMessageUnitTests(t *testing.T) {
+	resp, err := http.Get("https://raw.githubusercontent.com/simpleledger/slp-unit-test-data/master/script_tests.json")
```
Let's either hardcode these objects in the test or create a json file that is loaded for the test. 

We don't want this test to rely on the availability of that endpoint. Or even an internet connection for that matter.

jcramer replied:
done.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+	for _, test := range tests {
+		slpbuf, _ := hex.DecodeString(test.Script)
+		_, err := v1parser.ParseSLP(slpbuf)
+		if err != nil {
```
Probably should be rewritten as `if err != nil && test.code == nil {`

jcramer replied:
Resolved since TestSlpMessageUnitTests is in the goslp package unit tests.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+)
+
+// TestSlpMesesageUnitTests downloads SLP parser unit tests and checks the parser throws for each test where code != nil
+func TestSlpMessageUnitTests(t *testing.T) {
```
Is this test necessary? Seems to be testing SLP parsing but if it's being handled by the higher level library it seems this test should be in that library not BCHD right?

jcramer replied:
Ok, deleted since this test is in goslp package.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+
+// TestSlpInputUnitTests downloads SLP input unit tests and checks the input conditions for each test are met
+func TestSlpInputUnitTests(t *testing.T) {
+	resp, err := http.Get("https://raw.githubusercontent.com/simpleledger/slp-unit-test-data/master/tx_input_tests.json")
```
Same as above. Let's not use the internet to run the test.

jcramer replied:
done.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+	for _, test := range tests {
+
+		// create temporary db of input conditions
+		_entryDb := make(map[[32]byte]*SlpIndexEntry)
```
In Go we generally don't want variables starting with a _

jcramer replied:
done.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+			tx := wire.NewMsgTx(1)
+			serializedTx, err := hex.DecodeString(wen.Tx)
+			if err != nil {
+				panic(err.Error())
```
Use `t.Fatal(err)` here instead of a panic.

jcramer replied:
done.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+			}
+			tx.Deserialize(bytes.NewReader(serializedTx))
+			slpMsg, err := v1parser.ParseSLP(tx.TxOut[0].PkScript)
+			if err != nil || slpMsg == nil {
```
Are any errors supposed to happen here? If no, then we probably want to do `t.Error()` or `t.Fatal()`.

jcramer replied:
In this for loop we're accumulating valid inputs to use for checking against the unit test's "should" transaction.  If the input is invalid we can just skip it.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+			if err != nil || slpMsg == nil {
+				continue
+			}
+			_tokenid, err := goslp.GetSlpTokenID(tx)
```
leading _

jcramer replied:
done

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+				SlpVersionType: uint16(slpMsg.TokenType),
+				SlpOpReturn:    tx.TxOut[0].PkScript,
+			}
+			_hash := tx.TxHash()
```
leading _

jcramer replied:
done

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+		}
+
+		// add "When" and "Should" variables
+		_getSlpIndexEntry := func(txiHash *chainhash.Hash) (*SlpIndexEntry, error) {
```
leading _

jcramer replied:
done

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+			return slpEntry, nil
+		}
+
+		_putTxIndexEntry := func(tx *wire.MsgTx, slpMsg *v1parser.ParseResult, tokenIDHash *chainhash.Hash) error {
```
leading _

jcramer replied:
done

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+		if err != nil {
+			panic(err.Error())
+		}
+		tx.Deserialize(bytes.NewReader(serializedTx))
```
Should use `tx.BchDecode()` here instead of `tx.Deserialize()` as it's more generalized than deserialize.

jcramer replied:
done

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+
+		// check the slp txns
+		isValid, _ := CheckSlpTx(tx, _getSlpIndexEntry, _putTxIndexEntry)
+		if !isValid && !test.Should[0].Valid {
```
maybe rewrite as:
```
if isValid != test.Should[0].Valid {
    t.Errorf("Test %d: Expected valid = %t, got %t", i, test.Should[0].Valid, isValid)
}
```

jcramer replied:
good call!  done.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+		} else if isValid && test.Should[0].Valid {
+			continue
+		} else {
+			t.Fatal("input unit test failed")
```
Note that you only want to use `t.Fatal` in cases where the test is unrecoverable and cannot complete any other iterations. In this case it can continue on to other iterations so `t.Error` is more appropriate. 

Also, the error code of "input unit test failed" doesn't tell you what iteration failed. Putting some context in the error message like: ` t.Errorf("Test %d: Expected valid = %t, got %t", i, test.Should[0].Valid, isValid)` is more helpful in debugging.

jcramer replied:
good call! done.

cpacia commented on blockchain/indexers/slpindex_test.go:
```diff
@@ -0,0 +1,138 @@
+				continue
+			}
+			_tokenid, err := goslp.GetSlpTokenID(tx)
+			tokenIDHash, _ := chainhash.NewHash(_tokenid[:])
```
Should handle error here.

jcramer replied:
done.

cpacia commented on blockchain/indexers/manager.go:
```diff
@@ -47,10 +47,18 @@ func dbPutIndexerTip(dbTx database.Tx, idxKey []byte, hash *chainhash.Hash, heig
 // dbFetchIndexerTip uses an existing database transaction to retrieve the
 // hash and height of the current tip for the provided index.
-func dbFetchIndexerTip(dbTx database.Tx, idxKey []byte) (*chainhash.Hash, int32, error) {
+func dbFetchIndexerTip(dbTx database.Tx, indexerPtr *Indexer) (*chainhash.Hash, int32, error) {
```
Since `Indexer` is an interface this doesn't need to be a pointer.

jcramer replied:
thanks.  With the new `StartBlock()` method added, this code has been reverted to its original form.

cpacia commented on blockchain/indexers/manager.go:
```diff
@@ -47,10 +47,18 @@ func dbPutIndexerTip(dbTx database.Tx, idxKey []byte, hash *chainhash.Hash, heig
-func dbFetchIndexerTip(dbTx database.Tx, idxKey []byte) (*chainhash.Hash, int32, error) {
+func dbFetchIndexerTip(dbTx database.Tx, indexerPtr *Indexer) (*chainhash.Hash, int32, error) {
 	indexesBucket := dbTx.Metadata().Bucket(indexTipsBucketName)
+	indexer := *indexerPtr
```
This will be unnecessary after fixing the above.

jcramer replied:
done.

jcramer commented:
Note to self, consider returning list of tokenMetadata for getTransaction endpoint.

This would be more future proof and also allow burnt token metadata to be included.

zquestz commented on bchd.go:
```diff
@@ -148,6 +148,14 @@ func bchdMain(serverChan chan<- *server) error {
 
 		return nil
 	}
+	if cfg.DropSlpIndex {
```
NIT: I like to put a leading newline before conditionals.

zquestz commented on bchrpc/bchrpc.proto:
```diff
@@ -68,19 +70,44 @@ service bchrpc {
+
+    // GetTrustedValidation returns SLP validity related information for one or more transactions.
+    //
+    // **Requires SlpIndex with additional functionary key setup**
```
What is "functionary key setup"?

jcramer replied:
this was removed, please ignore

zquestz commented on bchrpc/bchrpc.proto:
```diff
@@ -102,6 +130,7 @@ service bchrpc {
     // grpc-web.
     //
     // **Requires TxIndex to receive input metadata**
+    // TODO: Add slp metadata similar to "SubscribeTransactions"
```
This still accurate? Should we expect it as part of this PR?

zquestz replied:
Did this happen?

jcramer replied:
I have removed the TODO note.  If the client wants token genesis metadata they can call GetTokenMetadata endpoint.

zquestz commented on bchrpc/bchrpc.proto:
```diff
@@ -472,6 +523,73 @@ message SubscribeBlocksRequest {
+        FunctionarySignature functionary_sig = 8;
+    }
+
+    // allows server to respond to a request with multiple
```
NIT: incomplete sentence.

zquestz replied:
Comment still not complete.

jcramer replied:
I will remove the comment, as this is obvious.

jcramer replied:
done

zquestz commented on bchrpc/bchrpc.proto:
```diff
@@ -683,7 +807,151 @@ message TransactionFilter {
+//
+message SlpToken {
+    bytes token_id = 1;
+    uint64 amount = 2 [jstype = JS_STRING];
```
+1 =)

zquestz commented on bchrpc/bchrpc.proto:
```diff
@@ -683,7 +807,151 @@ message TransactionFilter {
+    SlpVersionType version_type = 6;
+}
+
+message SlpTransactionInfo {
```
Should we include comments for these message types?

zquestz replied:
We should probably add a comment for this message type.

jcramer replied:
done.

zquestz commented on bchrpc/pb/bchrpc.pb.go:
```diff
@@ -1,7 +1,7 @@
 // versions:
-// 	protoc-gen-go v1.25.0
-// 	protoc        v3.12.3
+// 	protoc-gen-go v1.22.0
```
Bit worried about downgrading the versoin of protoc we are using to generate these. Can we bump protoc and regen?

jcramer replied:
done.

zquestz commented on bchrpc/proxy/.gitignore:
```diff
@@ -0,0 +1 @@
+gw
```
Since this file is nested in the repo, we should put this in the root .gitignore and remove this one.

zquestz replied:
Seems you missed a few comments from the previous review... most of my original comments had no response.

jcramer replied:
> Seems you missed a few comments from the previous review... most of my original comments had no response.

Correct, sorry about this.  It seems github had them collapsed and I probably needed to click "Show X more comments" to expand..  working on them now.

jcramer replied:
> Overall this looks great. Have we done IBD tests on testnet/mainnet? Just want to make sure that has been run on the latest code before we merge. =)

I will do this, hopefully will be completed by the end of the week.

zquestz commented on bchrpc/proxy/gen/.gitignore:
```diff
@@ -0,0 +1 @@
+*.go
```
Move to root .gitignore =)

jcramer replied:
done

zquestz commented on bchrpc/proxy/gw.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+var (
+	// command-line options:
```
NIT: These comments are self evident and don't provide a lot of value.

jcramer replied:
done.

zquestz commented on bchrpc/proxy/web/.gitignore:
```diff
@@ -0,0 +1 @@
+*.json
```
Move to root gitignore.

jcramer replied:
done.

zquestz commented on bchrpc/proxy/gw.go:
```diff
@@ -0,0 +1,61 @@
+		return err
+	}
+
+	// TODO: consider serving static files for swagger-ui
```
This would be very cool as a follow-up PR!

zquestz commented on bchrpc/server.go:
```diff
@@ -120,7 +127,8 @@ type GrpcServer struct {
 	ready    uint32 // atomic
-	shutdown int32  //atomic
+	shutdown int32  // atomic
+
```
NIT: Remove newline.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -411,7 +425,7 @@ func (s *GrpcServer) GetBlockInfo(ctx context.Context, req *pb.GetBlockInfoReque
 		h, err := chainhash.NewHash(req.GetHash())
 		if err != nil {
-			return nil, status.Error(codes.InvalidArgument, "invalid hash")
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid hash %s", err.Error()))
```
NIT: `fmt.Sprintf("invalid hash: %s", err.Error())`

jcramer replied:
I have found I can use `status.Errorf(codes.InvalidArgument, "invalid hash: %v", err)`.

Sound good?

zquestz commented on bchrpc/server.go:
```diff
@@ -448,7 +462,7 @@ func (s *GrpcServer) GetBlock(ctx context.Context, req *pb.GetBlockRequest) (*pb
 		h, err := chainhash.NewHash(req.GetHash())
 		if err != nil {
-			return nil, status.Error(codes.InvalidArgument, "invalid hash")
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid hash %s", err.Error()))
```
Same as above, I would add a `:` after invalid hash.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -527,7 +541,7 @@ func (s *GrpcServer) GetRawBlock(ctx context.Context, req *pb.GetRawBlockRequest
 		h, err := chainhash.NewHash(req.GetHash())
 		if err != nil {
-			return nil, status.Error(codes.InvalidArgument, "invalid hash")
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid hash %s", err.Error()))
```
A few more down here. Just is traditionally done as `error message: <val>` in error strings.

jcramer replied:
done, I think :)

zquestz commented on bchrpc/server.go:
```diff
@@ -640,18 +654,23 @@ func (s *GrpcServer) GetHeaders(ctx context.Context, req *pb.GetHeadersRequest)
 // GetTransaction returns a transaction given its hash.
 //
 // **Requires TxIndex**
+// **Requires SlpIndex for slp token metadata, and token metadata
```
NIT: Would reword as `**Requires SlpIndex for all token metadata`

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -670,8 +689,18 @@ func (s *GrpcServer) GetTransaction(ctx context.Context, req *pb.GetTransactionR
+			tokenID, _ := chainhash.NewHash(tx.SlpTransactionInfo.TokenId)
+			tokenMetadata, err = s.buildTokenMetadata(*tokenID)
+			if err != nil {
+				return nil, status.Error(codes.Internal, fmt.Sprintf("a unknown problem occured when building token metadata: %s", err.Error()))
```
Nice, this is how all the errors should be formatted. =)

zquestz commented on bchrpc/server.go:
```diff
@@ -670,8 +689,18 @@ func (s *GrpcServer) GetTransaction(ctx context.Context, req *pb.GetTransactionR
 
+		var tokenMetadata *pb.TokenMetadata
+		if req.IncludeTokenMetadata && tx.SlpTransactionInfo.ValidityJudgement == pb.SlpTransactionInfo_VALID {
+			tokenID, _ := chainhash.NewHash(tx.SlpTransactionInfo.TokenId)
```
Always make sure to handle errors, some are not obvious like not enough memory to allocate chainhash. =)

jcramer replied:
done

zquestz commented on bchrpc/server.go:
```diff
@@ -685,23 +714,33 @@ func (s *GrpcServer) GetTransaction(ctx context.Context, req *pb.GetTransactionR
 
+	var tokenMetadata *pb.TokenMetadata
+	if req.IncludeTokenMetadata && respTx.SlpTransactionInfo.ValidityJudgement == pb.SlpTransactionInfo_VALID {
+		tokenID, _ := chainhash.NewHash(respTx.SlpTransactionInfo.TokenId)
```
Handle error =)

jcramer replied:
done

zquestz commented on bchrpc/server.go:
```diff
@@ -759,7 +798,8 @@ func (s *GrpcServer) GetAddressTransactions(ctx context.Context, req *pb.GetAddr
 	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, "invalid address")
+		addr, err = goslp.DecodeAddress(req.Address, s.chainParams)
+		addr, err = goslp.ConvertSlpToCashAddress(addr, s.chainParams)
```
So I see you falling back to trying to decode the slp address and convert it to a cash address, however I don't see you handling the error case where both fail. If your fallback fails we should be returning an invalid address like before.

zquestz replied:
If we are doing this in any other spot we might want to write a custom decode function that encapsulates all this logic. It would make the error handling a bit prettier.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -844,7 +884,8 @@ func (s *GrpcServer) GetRawAddressTransactions(ctx context.Context, req *pb.GetR
 	addr, err := bchutil.DecodeAddress(req.Address, s.chainParams)
 	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, "invalid address")
+		addr, err = goslp.DecodeAddress(req.Address, s.chainParams)
```
Yeah we should have a high level function that takes care of this flow.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -983,26 +1039,44 @@ func (s *GrpcServer) GetAddressUnspentOutputs(ctx context.Context, req *pb.GetAd
+	if req.IncludeTokenMetadata && s.slpIndex != nil {
+		tokenMetadata = make([]*pb.TokenMetadata, 0)
+		for _hash := range tokenMetadataSet {
+			tm, _ := s.buildTokenMetadata(_hash)
```
We should have the error here and log it at the very least.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -983,26 +1039,44 @@ func (s *GrpcServer) GetAddressUnspentOutputs(ctx context.Context, req *pb.GetAd
 func (s *GrpcServer) GetUnspentOutput(ctx context.Context, req *pb.GetUnspentOutputRequest) (*pb.GetUnspentOutputResponse, error) {
+
+	if req.IncludeTokenMetadata && s.slpIndex == nil {
+		return nil, status.Error(codes.Unavailable, "slpindex required")
```
Nice, love the short circuit pattern here. =)

jcramer replied:
thanks :)

zquestz commented on bchrpc/server.go:
```diff
@@ -1020,6 +1094,12 @@ func (s *GrpcServer) GetUnspentOutput(ctx context.Context, req *pb.GetUnspentOut
 		coinbase = blockchain.IsCoinBase(tx)
+
+		// check if this txn is possibly an SLP transaction
+		_, err = v1parser.ParseSLP(tx.MsgTx().TxOut[0].PkScript)
```
Is there any chance of getting here with a tx with no outputs? Seems we should have a length check on the outputs.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1041,15 +1121,30 @@ func (s *GrpcServer) GetUnspentOutput(ctx context.Context, req *pb.GetUnspentOut
+		isSlpInMempool &&
+		req.IncludeMempool {
+		slpToken, err = s.getSlpToken(txnHash, req.Index)
+		tokenID, _ := chainhash.NewHash(slpToken.TokenId)
```
Error handling is good for your health. =)

jcramer replied:
done

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+		resp.ParsingError = er.Error()
+		return resp, nil
+	}
+	if slpMsg.TokenType == 0x41 && slpMsg.TransactionType == "GENESIS" {
```
Should we have a constant that represents this hard coded value? `0x41` is not going to be commonly known.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+		slpMsg, err := v1parser.ParseSLP(entry.SlpOpReturn)
+
+		// set the proper slp version type
+		if slpMsg.TokenType == 0x01 {
```
Yeah we definitely should have constants for each valid `TokenType`

jcramer replied:
done

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+
+		// set the proper slp version type
+		if slpMsg.TokenType == 0x01 {
+			if slpMsg.TransactionType == "SEND" {
```
I also would have constants for each `TransactionType`

jcramer replied:
done

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+		slpMsg, err := v1parser.ParseSLP(entry.SlpOpReturn)
+
+		// set the proper slp version type
+		if slpMsg.TokenType == 0x01 {
```
I would use a switch statement here.

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+			panic("trusted validation cannot return result for unknown slp version type")
+		}
+
+		if slpMsg.TransactionType == "SEND" {
```
I would use a switch here too...

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	pubKey, _ := extK.ECPubKey()
```
Errors! We just need to check all the spots. Because Go doesn't have exceptions, some of the errors that could be raised might surprise you!

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+func isMaybeSlpTransaction(txn *wire.MsgTx) bool {
+	if len(txn.TxOut) > 0 {
+		lokadHex, _ := hex.DecodeString("534c5000")
+		return bytes.Contains(txn.TxOut[0].PkScript, lokadHex)
```
Do we want to check that it contains or starts with?

jcramer replied:
We want contains because the scriptPubKey starts with OP_RETURN and one of a few kinds of data push opcodes.  Can be 0x04 or 0x4c04 etc.

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+func (s *GrpcServer) CheckSlpTransaction(ctx context.Context, req *pb.CheckSlpTransactionRequest) (*pb.CheckSlpTransactionResponse, error) {
+
+	if req.RequiredSlpBurns != nil {
+		return nil, status.Error(codes.Aborted, "intentional burning with 'AllowedSlpBurns' is not yet implemented")
```
Lets put in a TODO for this. =)

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+				continue
+			}
+			if !bytes.Equal(tokenIDHash, entry.TokenIDHash.CloneBytes()) {
+				// TODO: (BONUS) check 'req.RequiredSlpBurns' for allowed burns
```
Lets try to get this implemented. =)

jcramer replied:
will do, I have it partially implemented in my stash :)

zquestz commented on bchrpc/server.go:
```diff
@@ -1816,6 +2316,234 @@ func (s *GrpcServer) fetchTransactionsByAddress(addr bchutil.Address, startHeigh
+	slpMsg, _ := v1parser.ParseSLP(entry.SlpOpReturn)
+
+	// set isMintBaton and also check that vout is within proper range
+	if slpMsg.TransactionType == "MINT" {
```
NIT: Switch statement?

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1816,6 +2316,234 @@ func (s *GrpcServer) fetchTransactionsByAddress(addr bchutil.Address, startHeigh
+	return slpToken, nil
+}
+
+// manageSlpEntryCache keeps the SlpEntryCache updated on transaction and block events
```
NIT: Mention this is launched as a goroutine and does not return errors!

jcramer replied:
done.

zquestz commented on bchrpc/server.go:
```diff
@@ -1816,6 +2316,234 @@ func (s *GrpcServer) fetchTransactionsByAddress(addr bchutil.Address, startHeigh
+
+			err := s.slpIndex.AddMempoolTx(txDesc.Tx)
+			if err != nil {
+				continue
```
We should log failures here for tracking.

jcramer replied:
done.

zquestz commented on bchrpc/txfilter.go:
```diff
@@ -147,6 +165,7 @@ func (f *txFilter) AddRPCFilter(rpcFilter *pb.TransactionFilter, params *chaincf
 	for _, addrStr := range rpcFilter.GetAddresses() {
 		addr, err := bchutil.DecodeAddress(addrStr, params)
 		if err != nil {
+			// TODO: handle addresses in slpAddr format
```
This seems important to add =)

jcramer replied:
done.

zquestz commented on blockchain/indexers/slpindex.go:
```diff
@@ -0,0 +1,877 @@
+	)
+	if len(serializedData) == 70 {
+		if slpVersion == 0x41 {
+			panic("cannon have this length with nft1 child")
```
NIT: spelling. Also, we should not panic here or we will cause a very nasty unexpected shutdown, we need to make sure it just bubbles up errors and shuts down if needed.

jcramer replied:
done.

zquestz replied:
Still a panic? Maybe a commit is missing? Seems both panics are not fixed, but I see a done comment.

zquestz commented on blockchain/indexers/slpindex.go:
```diff
@@ -0,0 +1,877 @@
+	}
+
+	if metadata.NftGroupID == nil && metadata.SlpVersion == 0x41 {
+		panic("missing nft group id for NFT child " + string(id))
```
Return errors, no panics here!

jcramer replied:
done, also fixed other panics ;)

zquestz replied:
Still see a panic?

zquestz commented on blockchain/indexers/slpindex.go:
```diff
@@ -0,0 +1,877 @@
+// CheckSlpTx checks a transaction for validity and adds valid txns to the db
+func CheckSlpTx(tx *wire.MsgTx, getSlpIndexEntry GetSlpIndexEntryHandler, putTxIndexEntry AddTxIndexEntryHandler) (bool, []*BurnedInput) {
+
+	txSlpMsg, _ := v1parser.ParseSLP(tx.TxOut[0].PkScript)
```
NIT: do we need a length check on outputs?

jcramer replied:
done. Also handled other cases of the same thing.

zquestz commented on blockchain/indexers/slpindex.go:
```diff
@@ -0,0 +1,877 @@
+		return err
+	}
+
+	if slpMsg.TokenType != 0x01 && slpMsg.TokenType != 0x41 && slpMsg.TokenType != 0x81 {
```
Should we have an array or hash of accepted token types? Seems this logic is brittle and could easily be overlooked when adding a new type.

jcramer replied:
I added a new method:

```
func isAcceptableTokenVersionType(versionType int) bool {
	if versionType == v1parser.TokenTypeFungible01 ||
		versionType == v1parser.TokenTypeNft1Child41 ||
		versionType == v1parser.TokenTypeNft1Group81 {
		return true
	}
	return false
}

```

zquestz reviewed (changes requested):
Overall this looks great! There are a few things we really need to do to make sure this is ready to merge.

- Remove all magic values for `TokenType` and make them constants.
- Make sure all methods that can return an error can actually return the error and bubble it up.
- Add constants for `TransactionType` values.
- Format errors as `error text: <value>` for consistency.
- Add method to GoSLP package to handle the following pattern:

```
addr, err := bchutil.DecodeAddress(req.Address, s.chainParams)
	if err != nil {
		addr, err = goslp.DecodeAddress(req.Address, s.chainParams)
		addr, err = goslp.ConvertSlpToCashAddress(addr, s.chainParams)
         }
```
- If we decide not to return an error, we should at least log it in debug mode!
- Make sure anytime we access the first element of an array we have a length check before it to make sure the element exists!
- Use switch statements when all if/else conditions are looking at the same variable/value.
- Add comment to any methods only launched as a goroutine, and mention that it can't return errors!
- Run spell check on new comments! Most looked great but I spotted a few misspellings.

Let me know when you post changes, will try to review again more promptly. Thanks again for the awesome work on this!

jcramer commented on bchrpc/server.go:
```diff
@@ -929,6 +975,15 @@ func (s *GrpcServer) GetAddressUnspentOutputs(ctx context.Context, req *pb.GetAd
 
+			var slpToken *pb.SlpToken
+			if s.slpIndex != nil {
+				slpToken, _ = s.getSlpToken(&txHash, uint32(i))
```
For this particular case, if an error happens on line 980 then the condition on the following line will be skipped.  So I'm thinking it would be acceptable to leave this one as is.

zquestz commented:
Will try to review soon!

zquestz commented on bchrpc/txfilter.go:
```diff
@@ -207,6 +242,31 @@ func (f *txFilter) MatchAndUpdate(tx *bchutil.Tx, params *chaincfg.Params) bool
 	matched := f.matchAll
 
+	slpMsg, err := v1parser.ParseSLP(tx.MsgTx().TxOut[0].PkScript)
+	if err == nil {
```
NIT: I find it odd to check err == nil... is slpMsg nil if the parse fails? Can we test if slpMsg != nil?

jcramer replied:
fixed in 0aaf3ab, I also added a length check on `tx.MsgTx().TxOut`.

jcramer replied:
I also have a bunch of local updates for the burn allowance, hope to push here soon.

jcramer commented:
@zquestz @cpacia Please make your final reviews, all features and tests on my itinerary are completed.  I think I've addressed all of your comments :)

zquestz commented on bchrpc/bchrpc.proto:
```diff
@@ -683,7 +803,152 @@ message TransactionFilter {
-    // Subscribe/Unsubscribe to everything. Other filters
-    // will be ignored.
+    // Subscribed/Unsubscribe to everything. Other filters
+    // will be ignored.    
```
NIT: extra whitespace.

jcramer replied:
done

zquestz commented on bchrpc/bchrpc.proto:
```diff
@@ -683,7 +803,152 @@ message TransactionFilter {
+    bytes group_token_id = 1;
+}
+
+enum SlpAction {       // everything except "NON_SLP" could be annotated with burn_flags  
```
Would prefer these comments to be before the enum declaration.

jcramer replied:
done

zquestz commented:
Overall this looks great. Have we done IBD tests on testnet/mainnet? Just want to make sure that has been run on the latest code before we merge. =)

jcramer commented:
> Overall this looks great. Have we done IBD tests on testnet/mainnet? Just want to make sure that has been run on the latest code before we merge. =)

I will do this now, hopefully, will be done by the end of the week.

zquestz commented:
Kicked off another run of tests, seems like a dependency install issue popped up.

zquestz commented on blockchain/indexers/slpindex.go:
```diff
@@ -879,7 +879,7 @@ func DropSlpIndex(db database.DB, interrupt <-chan struct{}) error {
 	}
 
-	return dropIndex(db, slpIndexKey, slpIndexName, interrupt)
+	return nil
```
NIT: This function can just be:

```
func DropSlpIndex(db database.DB, interrupt <-chan struct{}) error {
  return dropIndex(db, slpIndexKey, slpIndexName, interrupt)
}
```

jcramer replied:
good call!!! I overlooked that.  And done.

jcramer commented:
### SLP Test Harness Checklist (at 626a86f):
- [x] passes all [message parsing tests](https://github.com/simpleledgerinc/goslp/blob/master/v1parser/v1_parser_test.go)
- [x] passes all [dagging tests](https://github.com/simpleledgerinc/bchd/blob/slp-index/blockchain/indexers/slpindex_test.go)
- [ ] passes all [slp-comps](https://github.com/simpleledgerinc/slp-comps) on-chain validity comparisons[1][2][3]
- [ ] passes all [burn prevention tests](https://github.com/simpleledgerinc/grpc-bchrpc-node/tree/master/test)[4]
- [x] passes [message fuzzing tests](https://github.com/simpleledgerinc/goslp/tree/master/fuzz); campaign has wide coverage with no unhandled exceptions[5]

Testing Notes:
1. slp-comps scanned up to block 6xx,xxx; analyzed 1,xxx,xxx valid and xx,xxx invalid SLP transactions for 3 independent validator implementations
2. slp-comps was run with [SLPDB](https://github.com/simpleledger/SLPDB) and [gs++](https://github.com/blockparty-sh/cpp_slp_graph_search) validators with no discrepancies found on BCH mainnet
3. There were four transactions found to be mismatching with SLPDB validator.  All of these txids are are valid, but SLPDB thinks they are invalid.  This is non-critical bug in SLPDB software related to 0 output for NFT child SEND transactions.
4. grpc-bchrpc-node tests were run with `SCAN_KNOWN_BURNS = true` using the known burn case files (i.e., `BURNED_*.txt` files) recorded from slp-comps output up to block 6xx,xxx including ~xx,xxx burn transactions, there files were commited to grpc-bchrpc-node in xxxxxx.
5. go-fuzz performed >20,000,000 executions with no crashes or unhandled exceptions using 12 cores for 2 hours.

zquestz commented:
> ### SLP Test Harness Checklist (at [626a86f](https://github.com/gcash/bchd/commit/626a86f7a80c3b6d8a6bfd44734d3d3541cf59f5)):
> * [x]  passes all [message parsing tests](https://github.com/simpleledgerinc/goslp/blob/master/v1parser/v1_parser_test.go)
> * [x]  passes all [dagging tests](https://github.com/simpleledgerinc/bchd/blob/slp-index/blockchain/indexers/slpindex_test.go)
> * [ ]  passes all [slp-comps](https://github.com/simpleledgerinc/slp-comps) on-chain validity comparisons[1][2][3]
> * [ ]  passes all [burn prevention tests](https://github.com/simpleledgerinc/grpc-bchrpc-node/tree/master/test)[4]
> * [x]  passes [message fuzzing tests](https://github.com/simpleledgerinc/goslp/tree/master/fuzz); campaign has wide coverage with no unhandled exceptions[5]
> 
> Testing Notes:
> 
> 1. slp-comps scanned up to block 6xx,xxx; analyzed 1,xxx,xxx valid and xx,xxx invalid SLP transactions for 3 independent validator implementations
> 2. slp-comps was run with [SLPDB](https://github.com/simpleledger/SLPDB) and [gs++](https://github.com/blockparty-sh/cpp_slp_graph_search) validators with no discrepancies found on BCH mainnet
> 3. There were four transactions found to be mismatching with SLPDB validator.  All of these txids are are valid, but SLPDB thinks they are invalid.  This is non-critical bug in SLPDB software related to 0 output for NFT child SEND transactions.
> 4. grpc-bchrpc-node tests were run with `SCAN_KNOWN_BURNS = true` using the known burn case files (i.e., `BURNED_*.txt` files) recorded from slp-comps output up to block 6xx,xxx including ~xx,xxx burn transactions, there files were commited to grpc-bchrpc-node in xxxxxx.
> 5. go-fuzz performed >20,000,000 executions with no crashes or unhandled exceptions using 12 cores for 2 hours.

This is amazing. Hopefully when this is done, and @cpacia finishes his remaining cleanup we can finally merge this puppy!

jcramer commented:
OK! All tests have passed @zquestz @cpacia

jcramer commented:
In the list above I've added a new test case for SLP validation WRT transactions in the mempool.  I will let you know when complete.

jcramer commented:
Sorry about the --force commits, I had a mistake in the graph search feature and wanted to replace it as a stand-alone feature commit.

jcramer commented:
I've removed the graph search feature commit because it just needs more time in testing and it adds a bit of complexity to the code.  

I can follow-up later with a subsequent PR for graph search feature.  Not many people would be using it either.

zquestz commented:
Glad to hear you are waiting on the graph search. We can add that in a later release with more testing/time. =)

zquestz commented:
So I see a few checkboxes above that aren't checked.

[ ] passes rpc api tests
[ ] passes all slp-comps on-chain validity comparisons[1][2][3]
[ ] passes all burn prevention tests[4]

Are we planning on resolving these before merge? Other than that this looks ready to go to me!

jcramer commented:
> Are we planning on resolving these before merge? Other than that this looks ready to go to me!

I will run the full suite of tests asap and circle back when ready.

jcramer commented:
### SLP Test Harness Checklist (at 626a86f):
- [ ] passes all [message parsing tests](https://github.com/simpleledgerinc/goslp/blob/master/v1parser/v1_parser_test.go)
- [ ] passes all [dagging tests](https://github.com/simpleledgerinc/bchd/blob/slp-index/blockchain/indexers/slpindex_test.go)
- [ ] passes all [slp-comps](https://github.com/simpleledgerinc/slp-comps) on-chain validity comparisons[1][2][3]
- [ ] passes all [burn prevention tests](https://github.com/simpleledgerinc/grpc-bchrpc-node/tree/master/test)[4]
- [ ] passes [message fuzzing tests](https://github.com/simpleledgerinc/goslp/tree/master/fuzz); the campaign has wide coverage with no unhandled exceptions[5]
- [ ] passes mempool caching[6]
- [ ] passes HTTP gRPC gateway api tests[7]

Testing Notes:
1. slp-comps scanned up to block 66_,___; analyzed over 1M possible SLP transactions for 3 independent validator implementations
2. slp-comps was run with [SLPDB](https://github.com/simpleledger/SLPDB) and [gs++](https://github.com/blockparty-sh/cpp_slp_graph_search) validators with no discrepancies found on BCH mainnet
3. There were four transactions found to be mismatched with SLPDB validator.  All of these transaction IDs are valid, but SLPDB thinks they are invalid.  This is a non-critical bug in SLPDB software related to 0 output for NFT child SEND transactions.
4. grpc-bchrpc-node was used to check the gRPC server's burn prevention logic. Tests in `test/client.spec.ts` were run with `SCAN_KNOWN_BURNS = true` using all 68,337 burn cases which are recorded up to block 66_,___. These files are commited in https://github.com/simpleledgerinc/grpc-bchrpc-node/commit/____.
5. go-fuzz performed an additional 78,000,000 parsing executions with no fuzzer crashes, no unhandled exceptions, using 6 CPUs for 7 hours.  This is on top of the existing test corpus which has over 2 weeks of fuzzer processing time logged.
6. Unconfirmed slp transaction metadata is available using a cache.  A manual test was performed to make sure only valid slp token transactions are able to be stored in this cache.  Acceptance into this cache occurs in `AddPotentialSlpMempoolTransaction()`.  Removal from this cache occurs on block acceptance events in `manageSlpEntryCache()`.
7. `./bchrpc/proxy/gw_test.go` is provides tests for the gRPC http gateway

cpacia commented on bchrpc/bchrpc.proto:
```diff
@@ -68,19 +70,44 @@ service bchrpc {
+    // GetBip44HdAddress returns an address associated with a provided xpub, address index, and
+    // selecting either internal or external by setting the change boolean.  If SlpIndex is disabled
+    // the slpAddress string will be an empty string.
+    rpc GetBip44HdAddress(GetBip44HdAddressRequest) returns (GetBip44HdAddressResponse) {}
```
Do we really need this one? It seems like this is just exposing a library function rather than accessing any unique data that the server has. Shouldn't this be done client side?

zquestz replied:
I don't mind this being there, but do agree with Chris's point. These types of functions seem better for bchutil, rather than the gRPC implementation.

jcramer replied:
Ok, will remove.

jcramer replied:
This is now removed.

cpacia commented:
Can you also run `go-bindata -pkg=main sample-bchd.conf` in the main directory. It will rebuild the bindata for the config file with the change you made to it.

jcramer commented on bchrpc/txfilter.go:
```diff
@@ -207,6 +241,33 @@ func (f *txFilter) MatchAndUpdate(tx *bchutil.Tx, params *chaincfg.Params) bool
 
+	if len(tx.MsgTx().TxOut) > 0 {
+		slpMsg, _ := v1parser.ParseSLP(tx.MsgTx().TxOut[0].PkScript)
+		if slpMsg != nil {
```
This was changed to use `err == nil && slpMsg != nil`.

jcramer commented:
> Can you also run `go-bindata -pkg=main sample-bchd.conf` in the main directory. It will rebuild the bindata for the config file with the change you made to it.

this is done.

jcramer commented:
I am currently investigating how to resolve returning  NFT Group ID for unconfirmed NFT transactions.  After this I will wrap up the final tests.

jcramer commented:
### SLP Test Harness Checklist (at commit ________):
- [ ] passes all [message parsing tests](https://github.com/simpleledgerinc/goslp/blob/master/v1parser/v1_parser_test.go)
- [ ] passes all [dagging tests](https://github.com/simpleledgerinc/bchd/blob/slp-index/blockchain/indexers/slpindex_test.go)
- [ ] passes all [slp-comps](https://github.com/simpleledgerinc/slp-comps) on-chain validity comparisons[1][2]
- [ ] passes all [burn prevention tests](https://github.com/simpleledgerinc/grpc-bchrpc-node/tree/master/test)[3]
- [ ] passes [message fuzzing tests](https://github.com/simpleledgerinc/goslp/tree/master/fuzz)[4]
- [ ] passes mempool caching[5]
- [ ] passes HTTP gRPC gateway API tests[6]
- [ ] passes re-indexing benchmark[7]

Testing Notes:
1. slp-comps scanned up to block 66_,___; analyzed over 1.3M possible SLP transactions that were compared with two independent validator implementations (i.e., [gs++](https://github.com/blockparty-sh/cpp_slp_graph_search) and [SLPDB](https://github.com/simpleledger/SLPDB)) with no discrepancies found on BCH mainnet
2. There were four transactions found to be mismatched with SLPDB validator.  All of these transaction IDs are valid, but SLPDB thinks they are invalid for the same reason.  The reason being a non-critical bug in SLPDB software which is related to related to 0 outputs in a NFT child SEND transactions.
3. grpc-bchrpc-node was used to check the gRPC server's burn prevention logic. Tests in `test/client.spec.ts` were run with `SCAN_KNOWN_BURNS = true` using all 6x,xxx burn cases which are recorded up to block 66_,___. These files are commited in https://github.com/simpleledgerinc/grpc-bchrpc-node/commit/____.
4. go-fuzz was used to perform an additional 78,000,000 parsing executions with no fuzzer crashes, no unhandled exceptions, using 6 CPUs for 7 hours.  This is on top of the existing test corpus which has over 2 weeks of fuzzer processing time logged.
5. Unconfirmed slp transaction metadata is available using a cache.  A manual test was performed to make sure only valid slp token transactions are able to be stored in this cache.  Acceptance into this cache occurs in `AddPotentialSlpMempoolTransaction()`.  Removal from this cache occurs on block acceptance events in `slpEventHandler()`.
6. `./bchrpc/proxy/gw_test.go` is provides tests for the gRPC http gateway
7. Time to index all slp transactions and metadata up to block 671,557 took 26 minutes and 17 minutes with SSD and NVMe drive types, respectively.  Re-indexing was accomplished  using `--dropslpindex` followed by `--slpindex`.

jcramer commented:
Just closed/opened to restart the Travis CI, the build failed for some network timeout on the last run.

jcramer commented:
My checks are complete using the latest commit 7035fa7.

jcramer commented:
![](https://gifimage.net/wp-content/uploads/2017/08/ron-paul-its-happening-gif-39.gif)
